### PR TITLE
Moved MurmurHash3 to external

### PR DESCRIPTION
**Pull request by dkeeney**
I set up a generic common.lib to contain all externals for which the sources are included within our repository (not downloaded).  The common.lib is merged into the static and dynamic nupic.core libraries.

The MurmurHash3 code is moved to external/common/murmurhash3.
Accessed via #include <murmurhash3/MurmurHash3.hpp>
linked via common.lib

### CODE REVIEW

File: external/common.cmake [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# -----------------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2016, Numenta, Inc.  Unless you have purchased from
+# Numenta, Inc. a separate commercial license for this software code, the
```
Comment (breznak) [inline external/common.cmake]: nit, to new files you don't have to add the numenta copyright, unless you want to ofc.

File: external/common.cmake [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+# that is physically included in this repository.
+#
+# Directions for adding a package:
+# 1) If it is an include only package, then just add it to external/include and skip the rest of these directions.
+# 2) Create a folder for the package in external/common/ and copy all files there.
+# 3) Add the .cpp files to the common_src list.
+# 4) The include files will be accessable with a path starting with the name of the folder created for your package.
+# 5) The objects will be placed in common.lib and included as part of the external libraries.
+#
+# Notes about MurmurHash3
```
Comment (breznak) [inline external/common.cmake]: nit, I'd add this text to murmur hpp file, not here

File: external/common/murmurhash3/MurmurHash3.cpp [hunk 3]
```diff
@@ -10,15 +10,15 @@ namespace nupic {
  * Don't worry about the technically undefined behavior when r >= 32, since this
  * is only used with a hardcoded r
  */
-inline UInt32 rotl32 ( UInt32 x, char r )
+inline unsigned int rotl32 ( unsigned int x, char r )
```
Comment (breznak) [inline external/common/murmurhash3/MurmurHash3.cpp]: this probably should be explicitely 32bit type, so unsigned int32_t 
https://en.cppreference.com/w/cpp/types/integer

File: external/common/murmurhash3/MurmurHash3.hpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+//      - removed all the other functions which we dont need.
+//      - added namespace nupic.
+//
+//    The seed argument type and the return type must match type UInt32.  This algorithm
+//    only works for 32bit integers.
+
+
+namespace nupic {
+
+  unsigned int MurmurHash3_x86_32( const void * key, int len, unsigned int seed );
```
Comment (breznak) [inline external/common/murmurhash3/MurmurHash3.hpp]: int32_t

File: external/common.cmake (hunk 1 above)
Comment (dkeeney) [reply external/common.cmake]: >  you don't have to add the numenta copyright, unless you want to ofc.

Everything I do is donated to Numenta.  I have no need to retain any rights.

File: external/common.cmake (hunk 2 above)
Comment (dkeeney) [reply external/common.cmake]: >  add this text to murmur hpp file,

It is in both places.

File: external/common/murmurhash3/MurmurHash3.cpp (hunk 3 above)
Comment (dkeeney) [reply external/common/murmurhash3/MurmurHash3.cpp]: > this probably should be explicitely 32bit type

yes, good catch.

**Comment by dkeeney** [comment]
Linux is not generating the common library.  Have not figured out why yet.

File: external/common/murmurhash3/MurmurHash3.hpp (hunk 4 above)
Comment (dkeeney) [reply external/common/murmurhash3/MurmurHash3.hpp]: > int32_t

Actually it is `uint32_t`.   It does not seem to like `unsigned int32_t`.

File: external/common.cmake (hunk 2 above)
Comment (dkeeney) [reply external/common.cmake]: > add this text to murmur hpp file,

It is in both places. But that might not be a good thing.  Someone might update the text in one place but not know about the other.   Removing it from common.cmake, leaving it only in MurmurHash3.hpp.

**Comment by dkeeney** [comment]
Thanks for the review.

**Pull request merged**